### DISCUSSION TITLE
Diagonal Cropping

### DISCUSSION THREAD

**Discussion by uriel-tannen** [ImageMagick/ImageMagick#2956] [Help] [answered]
I have a set of images such as the following, which I'm trying to split across the 45 degree diagonal into two separate images. So one that is grey, and another that is yellow/orange.

![diagonal](https://user-images.githubusercontent.com/36728824/101664359-a9262000-3a43-11eb-88c5-e01abfa95912.png)

I'm trying to adapt the following code, but am having issues (I'm using ImageMagick 7 on Windows). 

`magick convert *.jpg -gravity north -crop 145x145+0+0 +repage -draw "polygon 145,145 0,145 145,0 fill none alpha 135,135 floodfill" diagonal.png`

The error I'm getting is "non-conforming drawing primitive definition `alpha'"

It'd be great if I could get some assistance on how to solve what I'm trying to do...thank you in advance!

**Comment by fmw42** [comment, marked as answer]
Your image is only 40x40. So cropping at 145x145 is not going to be appropriate. What do you expect for your output?  It is also a png and not a jpg. Note that in IM 7, use magick not magick convert.

Perhaps what you want is to create a triangular binary mask and then put the mask into the alpha channel of the input


![img](https://user-images.githubusercontent.com/13458218/101670614-3cd9fb00-3a08-11eb-90a5-08e9d33f02db.png)

```
magick img.png -write mpr:img +delete \
\( mpr:img -fill white -colorize 100 -fill black -draw "polygon 40,40 0,40 40,0" \
mpr:img +swap -alpha off -compose copy_opacity -composite +write result1.png \) \
\( mpr:img -fill black -colorize 100 -fill white -draw "polygon 40,40 0,40 40,0" \
mpr:img +swap -alpha off -compose copy_opacity -composite +write result2.png \) \
null:

```

![result1](https://user-images.githubusercontent.com/13458218/101671070-def9e300-3a08-11eb-9619-1f7e952a65e1.png)

<br>

![result2](https://user-images.githubusercontent.com/13458218/101671072-df927980-3a08-11eb-8d94-f8a6992692a3.png)

**Reply by uriel-tannen** [reply]
Thanks @fmw42 - this is almost perfect! However for a reason I don't fully understand, the output images don't function well with a different script I have. The following code gives me the average RGB value for a given image:

`magick *.jpg -format "%f,%[fx:int(mean.r*255)],%[fx:int(mean.g*255)],%[fx:int(mean.b*255)]\n" info: > swatchesRGB.csv`

I replaced the *.jpg with *.png given the output image format (necessary for the transparent triangle), and it's giving me the same RGB(145,120,88) average value for both result1.png and result2.png. RGB(145,120,88) is actually the average colour of the original image, i.e. the mid-point between the grey and the brown - and not the average of the grey triangle alone or the average of the brown triangle alone.

I suspect this is down to me not properly understanding alpha channels (or something else), but it seems that the triangles retain their full and original pixel values despite their visible appearance. How do I change (either script) so that I can get the average colour of each of the (visible parts of the) triangular outputs?

Thanks once again.

**Reply by fmw42** [reply]
It works fine for me.

![img](https://user-images.githubusercontent.com/13458218/101687539-8124c580-3a1f-11eb-9e1f-5362693e2212.png)


```
magick img.png -scale 1x1! -alpha off -format "%[pixel:u.p]\n" info:
```

**srgb(56.8824%,47.2375%,34.6645%)**



```
magick img.png -write mpr:img +delete \
mpr:img -fill black -colorize 100 -fill white -draw "polygon 40,40 0,40 40,0" \
mpr:img +swap -alpha off -compose copy_opacity -composite +write result2.png \
-scale 1x1! -alpha off -format "%[pixel:u.p]\n" info:
```

![result2](https://user-images.githubusercontent.com/13458218/101687559-897d0080-3a1f-11eb-8773-ae06865c3444.png)

**srgb(62.15%,43.5525%,19.5538%)**

**Reply by uriel-tannen** [reply]
Fantastic, thanks @fmw42. I'll accept this answer.

**Reply by snibgo** [reply]
I'll add a comment:

"%[fx:int(mean.r)*255]" uses the overall mean of the red channel, ignoring alpha.

We can first use "-scale 1x1!" which makes one pixel, using colour values weighted by alpha, so transparent pixels are ignored. Then "%[fx:int(mean.r)*255]" will do what you want.

**Reply by fmw42** [reply]
Good point!

**Reply by uriel-tannen** [reply]
> I'll add a comment:
> 
> "%[fx:int(mean.r)*255]" uses the overall mean of the red channel, ignoring alpha.
> 
> We can first use "-scale 1x1!" which makes one pixel, using colour values weighted by alpha, so transparent pixels are ignored. Then "%[fx:int(mean.r)*255]" will do what you want.

I'm not sure I understand this...does "-scale 1x1!" somehow change it into a single-pixel image? If so how is "%[fx:int(mean.r)*255]" the mean of the red channel for all pixels of the original image? How does it distil a 40x40 triangle into a single pixel? Or have I misunderstood this, and instead its altering _each_ pixel and weighting it by alpha?

**Reply by snibgo** [reply]
."-scale 1x1! is a bit like "-resize 1x1!", changing each image to be 1 pixel. It is simpler than resize, taking an arithmetic mean that is weighted by the alpha of each input pixel. Following that with "%[fx:int(mean.r*255)]" will take the mean in the red channel of that single-pixel image. (The mean of a single pixel is of course simply that pixel.) So the command would be:
```
magick *.jpg -scale 1x1! -format "%f,%[fx:int(mean.r*255)],%[fx:int(mean.g*255)],%[fx:int(mean.b*255)]\n" info: > swatchesRGB.csv
```

**Comment by GeeMack** [comment]
With ImageMagick you can modify a diagonal half of any image by using the image, a modified copy of the image, and the necessary mask. Here's one way to do it with IMv7 on Windows 10...

    magick input.jpg -write mpr:img0 -modulate 100,0 -write mpr:img1 ^
       -sparse-color voronoi "%[fx:(w-h)/2],%[fx:(h-w)/2] black %[fx:(w+h)/2],%[fx:(h+w)/2] white" ^
        mpr:img0 mpr:img1 -reverse -composite output.png

That starts by reading in the original image and saving a copy to a memory register "mpr:img0". Next it de-saturates it to grayscale with "-modulate 100,0" and saves that to another memory register "mpr:img1".

After that it uses "-sparse-color" and some FX expressions to create a black and white mask divided diagonally with a top-left section and a bottom-right section.

Then it brings both the original image and the modified grayscale version back into the command, and reverses the order of the list to place the original, the modified version, and the mask in the correct order for compositing.

![sample1](https://user-images.githubusercontent.com/65475279/101686218-60b03780-3a2e-11eb-989b-b2f7ff031904.png)

Since IM's default behavior when compositing three images is to use the last one as a mask, a simple "-composite" creates the diagonally split output image.

![sample2](https://user-images.githubusercontent.com/65475279/101686273-745b9e00-3a2e-11eb-9ca9-874a8da9145e.png)

Those FX expressions will divide any image diagonally whether landscape or portrait orientation and regardless of aspect. You can change the direction of the diagonal by adding "-flop" immediately after creating the mask. You can change which corner gets the original and which gets the modified version by swapping the "black" and "white" in the "-sparse-color" operation. 

In your example it looks like you're reading in all the JPG images with "*.jpg". Since the above command uses a "-composite" operation it can be very tricky to manipulate  multiple input images. The better approach would be to run the command inside a "for" loop to step through all the JPG files.

This command works exactly the same with IMv6 using "convert" instead of "magick". I've also tested it with an old IMv6 on a bash shell on a *nix system by escaping the parentheses "\\(...\\)", changing the double-quotes to single quotes, and changing the continued-line carets "^" to backslashes "\\". There is a pretty limited use of FX expressions in IMv6, but "-sparse-color" is one of the operations where you can use them.

**Reply by uriel-tannen** [reply]
Thanks for this @GeeMack - Fred has already posted an answer that works for me, however I've also learned a fair bit from yours...so thanks for that!

**Reply by pbsds** [reply]
For those interested, I adapted this to combine two image files:
```shell
magick image_0.png -write mpr:img0 +delete image_1.png -write mpr:img1 \
    -sparse-color voronoi "%[fx:(w-h)/2],%[fx:(h-w)/2] white %[fx:(w+h)/2],%[fx:(h+w)/2] black" \
    mpr:img0 mpr:img1 -reverse -composite output.png
```

**Reply by fmw42** [reply]
An example would be useful.
